sdk/state/snapshot_test.go

```diff
@@ -0,0 +1,248 @@
+package state
+
+import (
+	"encoding/json"
+	"testing"
+	"time"
+
+	"github.com/stellar/experimental-payment-channels/sdk/txbuildtest"
+	"github.com/stellar/go/keypair"
+	"github.com/stellar/go/network"
+	"github.com/stellar/go/txnbuild"
+	"github.com/stellar/go/xdr"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func assertChannelSnapshotsAndRestores(t *testing.T, config Config, channel *Channel) {
+	t.Helper()
+
+	snapshot := channel.Snapshot()
+	snapshotJSON, err := json.Marshal(snapshot)
+	require.NoError(t, err)
+	restoredSnapshot := Snapshot{}
+	err = json.Unmarshal(snapshotJSON, &restoredSnapshot)
+	require.NoError(t, err)
+
+	restoredChannel := NewChannelWithSnapshot(config, snapshot)
+
+	require.Equal(t, channel, restoredChannel)
+}
+
+func TestNewChannelWithSnapshot(t *testing.T) {
+	localSigner := keypair.MustRandom()
+	remoteSigner := keypair.MustRandom()
+	localEscrowAccount := keypair.MustRandom().FromAddress()
+	remoteEscrowAccount := keypair.MustRandom().FromAddress()
+
+	localConfig := Config{
```

Comment:
❓  the current snapshot relies on the user to remember these fields in the `Config` struct, which then can be used to rehydrate a channel. But what if the user loses or accidentally changes these values? ie, why not include these in the snapshot?

Reply:
>what if the user loses or accidentally changes these values

Applications building on this should have tests that pick that up pretty quickly. It's the sort of thing if it is broken, it's always broken rather than intermittently broken.

>why not include these in the snapshot?

My first attempt at adding snapshotting had the function signature as `NewChannelFromSnapshot(s Snapshot)` without the config, however the `Config`s for both the `Channel` and the `Agent` both contain a mix of state information, dependencies (e.g. `Submitter`, `Streamer`, etc), and sensitive information (e.g. signing keys). If we include some of the `Config` fields in the `Snapshot` struct when rehydrating we have to handle the case where they're inconsistent, and what should happen then.

I think the simplest path at the moment is to say that the `Snapshot` contains anything about the internal state of the channel that isn't already captured in the `Config` that's required to rehydrate the channel.

I have some ideas for how to refactor `Config` so that it would be easier to move more into the `Snapshot` to be more like what you suggest, but it's a bigger piece of work and it'll be easier to see how effective the refactor is if the `Snapshot` is already in place before that happens.

Reply:
ok sgtm 👍 . Could you add this line: `Snapshot contains anything about the internal state of the channel that isn't already captured in the Config` as a comment above the Snapshot struct as a reminder?

Reply:
💯 Good call.  The docs are a bit thin and missing from some types/funcs I'm adding here.